### PR TITLE
Charts to make debugging easier

### LINKED ISSUE #31: Critic receives output from sensor instead of feature extractor
_compute_gae_jit: the output of the sensor is used as an input for the critic, while in update ppo, the PPO update uses the output of the feature extractor (the correct model), resulting in an out-of-distribution bug while training the brittle star

### PR DESCRIPTION
**Pull request by JibrilExe**
Explained variance, and terminated and truncated charts.

Closes #31

### CODE REVIEW

**Comment by JibrilExe** [comment]
Terminated/truncated over avg_ep_lengths feels redundant since it is just a combinated of what num_truncated/terminated combined with avg_ep_lengths tells us, feel free to tell me if you want it removed.

I assumed it was ok to add all new metrics to the LossInfo class, since it is still loss info related, but might be cleaner to change name to LogInfo since the losses are also mostly used in _log only.

**Comment by JibrilExe** [comment]
Example run: https://wandb.ai/SEL3-2026-Groep-4/LET-THERE-BE-MORE-LOGGING/runs/i2imn24c

File: scripts/hpc/install.sh [hunk 1]
```diff
@@ -46,6 +46,8 @@ source vsc-venv --activate \
 set -euo pipefail
 cd "$PBS_O_WORKDIR"
 
+uv run pip install .
+
```
Comment (tdpeuter) [inline scripts/hpc/install.sh]:
```suggestion
```

Please see `train.pbs` on how to fix this when running non-interactively on a cluster, by modifying `PYTHONPATH`  ([commit](https://github.com/SELab-3-2026/SEL3-2026-Groep-4/commit/3c6eec2410f7836cbef144dbaf69151971dcd6a0)). For interactive environments, I also [modified](https://github.com/SELab-3-2026/SEL3-2026-Groep-4/commit/f6dc9c8e7fcfe7c16a7c6f48096a7edd3e18198e) `pyproject.toml` such that the packages are picked up by `uv`.

If you added this line specifically for interactive sessions on the HPC, I think it is best to run the command interactively, but I can see how that might be frustrating. I'll have to come up with something. 

Can you verify this line can be removed?

File: src/brittle_star_project/trainers/PPOTrainer.py [hunk 2]
```diff
@@ -397,6 +414,23 @@ def _step(self, env_state, next_obs, next_done, iteration: int) -> tuple:
...
         )
 
+        explained_var = _compute_explained_variance(storage.values, storage.returns)
+
+        terminated = next_env_state.terminated  # (num_envs,)
+        truncated = next_env_state.truncated  # (num_envs,)
+        episode_lengths = self.episode_stats.returned_episode_lengths
+
+        num_terminated = int(jnp.sum(terminated).item())
+        num_truncated = int(jnp.sum(truncated).item())
```
Comment (tdpeuter) [inline src/brittle_star_project/trainers/PPOTrainer.py]: Does `terminated` represent in how many environments the robot has reached the target?

What does `truncated` represent?

File: src/brittle_star_project/trainers/PPOTrainer.py [hunk 3]
```diff
@@ -197,6 +203,12 @@ class LossInfo:
     entropy_loss: Any
     approx_kl: Any
     avg_episodic_return: Any
+    # Example of better typing
+    explained_variance: float
+    num_terminated: int
+    num_truncated: int
+    avg_terminated_length: Any
+    avg_truncated_length: Any
```
Comment (tdpeuter) [inline src/brittle_star_project/trainers/PPOTrainer.py]: Can we improve the typing of the other values as well, and clean up tgr comments, please? I would have to look at the code.

**Review by tdpeuter — CHANGES_REQUESTED** [review]
Thanks, Explained Variance is surprisingly easy to understand, or maybe, ... *explain*... *evil laugh*

I added some suggestions for cleaning up.

PS; don't forget to configure your git on the HPC when you want to make commits there.

**Comment by tdpeuter** [comment]
> Terminated/truncated over avg_ep_lengths feels redundant since it is just a combinated of what num_truncated/terminated combined with avg_ep_lengths tells us, feel free to tell me if you want it removed.

For now I think it's fine. We can always filter in our results as well, right?

> I assumed it was ok to add all new metrics to the LossInfo class, since it is still loss info related, but might be cleaner to change name to LogInfo since the losses are also mostly used in _log only.

That's okay. I think an even better name might be `Measurements` or something similar.

**Comment by tdpeuter** [comment]
> Closes #31 

I don't see any changes related to that issue. Can you verify the fixed issue, please? Did you still have local changes?

File: src/brittle_star_project/trainers/PPOTrainer.py (hunk 2 above)
Comment (JibrilExe) [reply src/brittle_star_project/trainers/PPOTrainer.py]: https://github.com/Co-Evolve/SEL3-tutorials-2026/blob/main/tutorials/brb_brittle_star_environment_tutorial.ipynb

Terminated means the goal was reached indeed, and truncated means the time limit was reached.
My goal of these charts is to get an idea of when envs truncate.

File: scripts/hpc/install.sh (hunk 1 above)
Comment (JibrilExe) [reply scripts/hpc/install.sh]: Ah ye it works, just verified by fully removing the venv and re-running again. And ah no it is not that frustrating, I added it just because I wanted to test on a different node so needed a new venv, and did not remember that you already fixed it, cheers.

**Review by RobinMeersman — APPROVED** [review]

**Review by tdpeuter — APPROVED** [review]
Thanks
